File: 08/day_eight.py


digit_to_segments = {
    0: [0,1,2,4,5,6],
    1: [2,5],
    2: [0,2,3,4,6],
    3: [0,2,3,5,6],
    4: [1,2,3,5],
    5: [0,1,3,5,6],
    6: [0,1,3,4,5,6],
    7: [0,2,5],
    8: [0,1,2,3,4,5,6],
    9: [0,1,2,3,5,6]
}


segment_count_to_digits = {
    2: [1],
    3: [7],
    4: [4],
    5: [2,3,5],
    6: [0,6,9],
    7: [8]
}


def get_patterns_and_outputs(line: str):
    elements = line.split()
    patterns = elements[:len(elements) - 5]
    outputs = elements[len(elements) - 4:]
    return patterns, outputs


def find_candidate_segments(char, patterns):

    pattern_segments = []
    containing_char = [p for p in patterns if char in p]
    for pattern in containing_char:
        possible_digits = segment_count_to_digits[len(pattern)]
        for digit in possible_digits:
            segments = digit_to_segments[digit]
            pattern_segments.append(segments)
    return min(pattern_segments, key=len)



def find_mapping(patterns, mapping):

    unmapped_chars = [c for c in mapping if mapping[c] is None]
    available_segments = [i for i in [0,1,2,3,4,5,6] if i not in mapping.values()]

    for char in unmapped_chars:
        candidate_segments = [s for s in find_candidate_segments(char, patterns) if s in available_segments]
        if not candidate_segments:
            raise ValueError("No available segments")
        for segment in candidate_segments:
            mapping[char] = segment
            try:
                find_mapping(patterns, mapping)
                try:
                    for pattern in patterns:
                        map_pattern(mapping, pattern)
                except KeyError:
                    raise ValueError("Mapping is not valid")
                return mapping
            except ValueError:
                mapping[char] = None
                continue
        else:
            raise ValueError("None of the available segments leads to a solution")

    return mapping


def map_pattern(mapping, pattern):

    segments = sorted([mapping[c] for c in pattern])
    digit = list(digit_to_segments.keys())[list(digit_to_segments.values()).index(segments)]
    return digit


def map_output_values(mapping, outputs):

    output_string = ""
    for output in outputs:
        digit = map_pattern(mapping, output)
        output_string += str(digit)
    return int(output_string)


def main():

    with open("data.txt") as data:
        lines = [l.replace("\n", "") for l in data.readlines()]

    output_numbers = []
    for line in lines:
        mapping = {c: None for c in "abcdefg"}
        patterns, outputs = get_patterns_and_outputs(line=line)
        mapping = find_mapping(patterns, mapping)
        output_numbers.append(map_output_values(mapping, outputs))

    print(sum(output_numbers))
    

if __name__ == "__main__":
    main()

File: 08/DayEight.hs
import System.IO
import Control.Monad
import Data.Map (Map, fromList, findWithDefault)
import Data.List (nub, intersect, sort, (\\))
import Data.Maybe (isNothing)
import Debug.Trace


type ConnectionMap = Map Char Int


digitToSegments :: Map Int [Int]
digitToSegments = fromList [
    (0, [0,1,2,4,5,6]),
    (1, [2,5]),
    (2, [0,2,3,4,6]),
    (3, [0,2,3,5,6]),
    (4, [1,2,3,5]),
    (5, [0,1,3,5,6]),
    (6, [0,1,3,4,5,6]),
    (7, [0,2,5]),
    (8, [0,1,2,3,4,5,6]),
    (9, [0,1,2,3,5,6])
    ]


segmentCountToDigits :: Map Int [Int]
segmentCountToDigits = fromList [
    (2, [1]),
    (3, [7]),
    (4, [4]),
    (5, [2,3,5]),
    (6, [0,6,9]),
    (7, [8])
    ]


getOutputs :: [[String]] -> [[String]]
getOutputs = map (\d -> drop (length d - 4) d)


countOutputDigits :: [String] -> Int  -- part one
countOutputDigits [] = 0
countOutputDigits outputs@(o:os)
    | length o `elem` [2,3,4,7] = 1 + countOutputDigits os
    | otherwise = countOutputDigits os


getCandidateSegmentsFromLength :: Int -> [Int]
getCandidateSegmentsFromLength l =
    let
        possibleDigits = findWithDefault [] l segmentCountToDigits
    in  nub $ concat $ map (\d -> findWithDefault [] d digitToSegments) possibleDigits


findCandidateSegments :: Char -> [String] -> [Int]
findCandidateSegments _ [] = [0..6]
findCandidateSegments c patterns@(p:ps)
    | c `elem` p = sort $ candidateSegments `intersect` findCandidateSegments c ps
    | otherwise = findCandidateSegments c ps
    where 
        candidateSegments = getCandidateSegmentsFromLength (length p)


findMapping :: [Char] -> [Int] -> [String] -> Int -> Maybe [(Char, Int)]
findMapping [] _ _ _ = Just []
findMapping _ [] _ _ = Nothing
findMapping _ _ [] _ = Nothing
findMapping chars@(c:cs) availableSegments patterns idx
    | null candidateSegments = Nothing
    | isNothing nextMapping = findMapping chars availableSegments patterns (idx + 1)
    | otherwise = fmap ((:) (c, segment)) (findMapping cs (availableSegments \\ [segment]) patterns idx)
    where
        candidateSegments = (findCandidateSegments c patterns) `intersect` availableSegments
        segment = candidateSegments !! idx
        nextMapping = findMapping cs (availableSegments \\ [segment]) patterns idx
        

main = do
    contents <- readFile "test.txt"
    let inputs :: [[String]] = map words $ lines contents
    let patterns :: [[String]] = map (\c -> take (length c - 5) c) inputs
    let outputs :: [[String]] = getOutputs inputs
    -- print $ countOutputDigits $ concat outputs
    print $ map (\c -> findCandidateSegments c (patterns !! 1)) "abcdefg"
    print $ findMapping "abcdefg" [0..6] (patterns !! 1) 0
